lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp

```diff
@@ -657,12 +657,21 @@ GetCanonicalNode(SwiftASTContext *module_holder,
 
     case Node::Kind::BoundGenericTypeAlias:
     case Node::Kind::TypeAlias: {
-      auto node_clangtype = ResolveTypeAlias(module_holder, dem, node);
-      if (CompilerType clang_type = node_clangtype.second)
-        return GetClangTypeNode(clang_type, dem, module_holder);
-      if (node_clangtype.first)
-        return node_clangtype.first;
-      return node;
+      while (true) {
+        auto pair = ResolveTypeAlias(module_holder, dem, node);
+        if (auto clang_type = std::get<CompilerType>(pair))
+          return GetClangTypeNode(clang_type, dem, module_holder);
+        if (NodePointer swift_node = std::get<NodePointer>(pair)) {
+          // A swift typealias may resolve to a clang typealias, which we must
+          // then resolve.
+          if (swift_node && swift_node->getKind() == Node::Kind::TypeAlias)
+            node = swift_node;
+          else
+            return swift_node;
```

Comment:
Could you add some sanity-check to avoid an infinite loop here if the typedefs end up being circular? Either a DenseSet of visited nodes or just a max_depth counter?

Reply:
maybe:
```
diff --git a/lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp b/lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp
index f651879a6e9d..2e7b3528c083 100644
--- a/lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp
+++ b/lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp
@@ -393,7 +393,12 @@ static std::pair<swift::Demangle::NodePointer, CompilerType>
 ResolveTypeAlias(SwiftASTContext *module_holder,
                  swift::Demangle::Demangler &dem,
                  swift::Demangle::NodePointer node,
-                 bool prefer_clang_types = false) {
+                 bool prefer_clang_types = false,
+                 unsigned char iter = 0) {
+  if (iter > 99)
+    LLDB_LOGF(GetLogIfAllCategoriesSet(LIBLLDB_LOG_TYPES),
+              "suspiciously long type alias chain, giving up");
+    return {{}, {}};
   LLDB_SCOPED_TIMER();
   auto resolve_clang_type = [&]() -> CompilerType {
     auto maybe_module_and_type_names = GetNominal(dem, node);
@@ -481,6 +486,10 @@ ResolveTypeAlias(SwiftASTContext *module_holder,
              "Unrecognized demangling %s", desugared_name.AsCString());
     return {{}, {}};
   }
+  // Resolve type aliases that are type aliases.
+  if (n->getKind() == Node::Kind::TypeAlias)
+    return ResolveTypeAlias(module_holder, dem, n, prefer_clang_types,
+                            iter + 1);
   return {n, {}};
 }
 ```
For some reason this alternative patch seems to introduce a ton of new test failures.

Reply:
When I applied my alternate patch I posted above yesterday it introduced a bunch of new test failure. Either means my patch has a bug, or it has unintended consequences because it run on every ResolveTypeAlias invocations. (Yours is more targeted).